### DISCUSSION TITLE
Adding listener when getting bucket through Supplier

### DISCUSSION THREAD

**Discussion by haiderali22** [vladimir-bukhtoyarov/bucket4j#184] [Q&A] [answered]
I am trying to figure out how to add my listener when bucket is created first time. I am using below code

`private Supplier<BucketConfiguration> getConfigSupplierForUser(String apiKey) {
		return () -> {
			return apiUsagePlanRepository.findByToken(apiKey).map( p -> {
				long translationsPerDay = 	p.getLimit();

				return Bucket4j.configurationBuilder()
						.addLimit(Bandwidth.simple(translationsPerDay, Duration.ofDays(1)))
						.build();
			}).orElseThrow( () -> {
				throw new IllegalArgumentException("API key not found");
			}  );

		};
	}
Supplier<BucketConfiguration> configurationLazySupplier = getConfigSupplierForUser(apiKey);

		// acquire cheap proxy to bucket  
		Bucket tokenBucket = buckets.getProxy(apiKey, configurationLazySupplier);

`
If i add listener to last line e.g. tokenBucket.toListenable(new MyListener()) then i think i am adding multiple listener to bucket
so i think there should be an option to add listener when creating BucketConfiguration

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
@haiderali22 Hello,

```java
// acquire cheap proxy to bucket  
Bucket tokenBucket = buckets.getProxy(apiKey, configurationLazySupplier);
```
The code above returns the bucket without listeners independently whether or not bucket had been created previously and listener were attached,  because listeners are never stored in the GRID. Bucket persists only configuration and currently available tokens, so you can safely modify code like bellow, there is no chance that it will be multiple listeners on the same bucket
```java
// acquire cheap proxy to bucket  
	Bucket tokenBucket = buckets.getProxy(apiKey, configurationLazySupplier).toListenable(myListener);
```
